### PR TITLE
`py-dm-tree`: add v0.1.8

### PR DESCRIPTION
**Pull request by thomas-bouvier**

### CODE REVIEW

**Review by manuelakuhn — COMMENTED** [review]
`py-six` is not required anymore since version 0.1.7

The `bazel` build dependency is not needed anymore for version 0.1.8. It builds fine without it (with the changes described below).

This package does not build for me (also version 0.1.6) throwing the error `error: legacy-install-failure` (same as the failed one in the pipeline).
~~Updating pip (see #38608) and~~ adding a missing `cmake` build dependency to prevent
```
FileNotFoundError: [Errno 2] No such file or directory: 'cmake'
```
did it for me and the version 0.1.8 installs with `--test=root` on Fedora37 and gcc@12.3.1

In addition the old versions don't build anymore with the current bazel version (after the changes above):
```
4 errors found in build log:
     50      Extracting Bazel installation...
     51      Starting local Bazel server and connecting to it...
     52      Loading:
     53      Loading:
     54      Loading: 0 packages loaded
     55      Analyzing: target //tree:_tree (1 packages loaded, 0 targets configured)
  >> 56    ERROR: /tmp/spack8vjlofad/772f85b7ccce5505562ee25453ff3a21/external/bazel_tools/platforms/BUILD:59:6: i
           n alias rule @@bazel_tools//platforms:osx: Constraints from @@bazel_tools//platforms have been removed. P
           lease use constraints from @@platforms repository embedded in Bazel, or preferably declare dependency on
            https://github.com/bazelbuild/platforms. See https://github.com/bazelbuild/bazel/issues/8622 for detai
           ls.
  >> 57      ERROR: /tmp/spack8vjlofad/772f85b7ccce5505562ee25453ff3a21/external/bazel_tools/platforms/BUILD:59:6:
            Analysis of target '@@bazel_tools//platforms:osx' failed
  >> 58      ERROR: /home/nela/projects/manuelakuhn/spack-stage/spack-stage-py-dm-tree-0.1.6-jyyonsv24inlpupin7smc
           xosfq4sx6ob/spack-src/tree/BUILD:27:18: errors encountered resolving select() keys for //tree:_tree
     59      ERROR: Analysis of target '//tree:_tree' failed; build aborted:
     60      INFO: Elapsed time: 2.079s
     61      INFO: 0 processes.
  >> 62      FAILED: Build did NOT complete successfully (6 packages loaded, 7 targets configured)
     63      error: command '/home/nela/projects/manuelakuhn/spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/baze
           l-6.1.1-oiytb2daurwiovn7pswfsfr63d3rrr2u/bin/bazel' failed with exit code 1
     64      error: subprocess-exited-with-error
     65    
     66      × python setup.py bdist_wheel did not run successfully.
     67      │ exit code: 1
     68      ╰─> See above for output.
```
I did a test with bazel@5.4.0 and there it built, so I guess a restriction on the bazel version is needed as well.

Tl;dr:
~~PR #38608 together with~~ this did the trick for me:
```
    depends_on("cmake", when="@0.1.8:", type="build")

    # Historical dependencies
    depends_on("bazel@:5", when="@:0.1.6", type="build")
    depends_on("py-six@1.12.0:", when="@:0.1.6", type=("build", "run"))
```

**Comment by manuelakuhn** [comment]
@adamjstewart do you know what changed in spack that we now get `legacy-install-failure` even for old versions of packages (here even when using old python versions to build)?

**Comment by manuelakuhn** [comment]
Correction: the update of pip ist not needed, that was a mistake on my side. Sorry for that.
But the rest is still true.

**Comment by adamjstewart** [comment]
^ @pradyunsg any ideas? I haven't personally seen the reported `legacy-install-failure` error but I trust @manuelakuhn

**Comment by thomas-bouvier** [comment]
I couldn't build version 0.1.6 neither, same `legacy-install-failure` error. No problem with version 0.1.8 though.

File: var/spack/repos/builtin/packages/py-dm-tree/package.py [hunk 1]
```diff
@@ -19,12 +19,15 @@ class PyDmTree(PythonPackage):
 
     maintainers("aweits")
 
+    version("0.1.8", sha256="0fcaabbb14e7980377439e7140bd05552739ca5e515ecb3119f234acee4b9430")
+    version("0.1.7", sha256="a54a5794b51ba585dc62f5d19c9d907f73107d5f0569417f8866ef3f75be5134")
```
Comment (manuelakuhn) [inline var/spack/repos/builtin/packages/py-dm-tree/package.py]: The checksum from github is different from that of pypi. Installation failed for me with:
```
==> Error: ChecksumError: sha256 checksum failed for $spack-stage/spack-stage-py-dm-tree-0.1.7-c4j6bj7p3parivvulvxrrip7ka23gbly/dm-tree-0.1.7.tar.gz
    Expected a54a5794b51ba585dc62f5d19c9d907f73107d5f0569417f8866ef3f75be5134 but got 30fec8aca5b92823c0e796a2f33b875b4dccd470b57e91e6c542405c5f77fd2a. File size = 35083 bytes. Contents = b'\x1f\x8b\x08\x08d\xeaSb\x02\xffdm-tre...\xc3\xcf\xe1\xe71\x7f\xfe\x1f\x81\xcb.\xbd\x00\xf8\x02\x00'
```

**Review by manuelakuhn — COMMENTED** [review]
With your changes now, version 0.1.8, 0.1.6 and 0.1.5 install for me with `--test=root`

**Comment by manuelakuhn** [comment]
> ^ @pradyunsg any ideas? I haven't personally seen the reported `legacy-install-failure` error but I trust @manuelakuhn

Maybe to give a little bit more details about the `legacy-install-failure`:

The very first time I encountered this error was in #38585 for py-ruamel-yaml-clib@0.2.0 when building with python@3.10. The fix was to use an older python version for this older py-ruamel-yaml-clib.

<details>
  <summary>Full error message of `py-ruamel-yaml-clib` with python@3.10</summary>

```
==> Installing py-ruamel-yaml-clib-0.2.0-hdspcqxwznokaxubih4nb37ksuocgnpr [25/25]
==> No binary for py-ruamel-yaml-clib-0.2.0-hdspcqxwznokaxubih4nb37ksuocgnpr found: installing from source
==> Using cached archive: $spack/var/spack/cache/_source-cache/archive/b6/b66832ea8077d9b3f6e311c4a53d06273db5dc2db6e8a908550f3c14d67e718c.tar.gz
==> No patches needed for py-ruamel-yaml-clib
==> py-ruamel-yaml-clib: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/python-3.10.10-hv6jvlmupid2ylf5gt2m6ahtojappsuf/bin/python3.10' '-m' 'pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-ruamel-yaml-clib-0.2.0-hdspcqxwznokaxubih4nb37ksuocgnpr' '.'

14 errors found in build log:
     12      Processing $spack-stage/spack-stage-py-ruamel-yaml-clib-0.2.0-hdspcqxw
             znokaxubih4nb37ksuocgnpr/spack-src
     13        Added file://$spack-stage/spack-stage-py-ruamel-yaml-clib-0.2.0-hdsp
             cqxwznokaxubih4nb37ksuocgnpr/spack-src to build tracker '/tmp/pip-build-tracker-frl_u51z'
     14        Running setup.py (path:$spack-stage/spack-stage-py-ruamel-yaml-clib-
             0.2.0-hdspcqxwznokaxubih4nb37ksuocgnpr/spack-src/setup.py) egg_info for package from file:/$spack-
			 stage/spack-stage-py-ruamel-yaml-clib-0.2.0-hdspcqxwznokaxubih4nb37ksuoc
             gnpr/spack-src
     15        Created temporary directory: /tmp/pip-pip-egg-info-ww6ijw5j
     16        Preparing metadata (setup.py): started
     17        Running command python setup.py egg_info
  >> 18      $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-setuptools-67.6.0
             -dytmea7woljpe5ma24z6of5lhwfe36r6/lib/python3.10/site-packages/setuptools/dist.py:287: SetuptoolsDepr
             ecationWarning: The namespace_packages parameter is deprecated, consider using implicit namespaces in
             stead (PEP 420). See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-
             packages
     19          warnings.warn(msg, SetuptoolsDeprecationWarning)
     20        sys.argv ['$spack-stage/spack-stage-py-ruamel-yaml-clib-0.2.0-hdspcq
             xwznokaxubih4nb37ksuocgnpr/spack-src/setup.py', 'egg_info', '--egg-base', '/tmp/pip-pip-egg-info-ww6i
             jw5j']
     21        test compiling test_ruamel_yaml
     22        running egg_info
     23        creating /tmp/pip-pip-egg-info-ww6ijw5j/ruamel.yaml.clib.egg-info
     24        writing /tmp/pip-pip-egg-info-ww6ijw5j/ruamel.yaml.clib.egg-info/PKG-INFO

     ...

     38        Created temporary directory: /tmp/pip-wheel-el66f1y9
     39        Building wheel for ruamel.yaml.clib (setup.py): started
     40        Destination directory: /tmp/pip-wheel-el66f1y9
     41        Running command python setup.py bdist_wheel
     42        sys.argv ['$spack-stage/spack-stage-py-ruamel-yaml-clib-0.2.0-hdspcq
             xwznokaxubih4nb37ksuocgnpr/spack-src/setup.py', 'bdist_wheel', '-d', '/tmp/pip-wheel-el66f1y9']
     43        test compiling test_ruamel_yaml
  >> 44      $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-setuptools-67.6.0
             -dytmea7woljpe5ma24z6of5lhwfe36r6/lib/python3.10/site-packages/setuptools/dist.py:287: SetuptoolsDepr
             ecationWarning: The namespace_packages parameter is deprecated, consider using implicit namespaces in
             stead (PEP 420). See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-
             packages
     45          warnings.warn(msg, SetuptoolsDeprecationWarning)
     46        running bdist_wheel
     47        running build
     48        running build_py
     49        creating build
     50        creating build/lib.linux-x86_64-cpython-310

     ...

     456             |                                                                                           |
     457             |                                                                                           c
             har *
     458       yaml.h:636:43: note: expected 'yaml_char_t *' {aka 'unsigned char *'} but argument is of type 'char
              *'
     459         636 |         yaml_char_t *anchor, yaml_char_t *tag, int implicit,
     460             |                              ~~~~~~~~~~~~~^~~
     461       _ruamel_yaml.c: In function '__pyx_tp_dealloc_12_ruamel_yaml_CParser':
  >> 462     _ruamel_yaml.c:23880:5: error: lvalue required as increment operand
     463       23880 |     ++Py_REFCNT(o);
     464             |     ^~
  >> 465       _ruamel_yaml.c:23882:5: error: lvalue required as decrement operand
     466       23882 |     --Py_REFCNT(o);
     467             |     ^~
     468       _ruamel_yaml.c: In function '__pyx_tp_dealloc_12_ruamel_yaml_CEmitter':
  >> 469       _ruamel_yaml.c:24053:5: error: lvalue required as increment operand
     470       24053 |     ++Py_REFCNT(o);
     471             |     ^~
  >> 472       _ruamel_yaml.c:24055:5: error: lvalue required as decrement operand
     473       24055 |     --Py_REFCNT(o);
     474             |     ^~
     475       _ruamel_yaml.c: In function '__Pyx_ParseOptionalKeywords':
     476       _ruamel_yaml.c:26224:21: warning: '_PyUnicode_get_wstr_length' is deprecated [-Wdeprecated-declarat
             ions]
     477       26224 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :
     478             |                     ^

     ...

     594       Building wheel for ruamel.yaml.clib (setup.py): finished with status 'error'
     595       ERROR: Failed building wheel for ruamel.yaml.clib
     596       Running setup.py clean for ruamel.yaml.clib
     597       Running command python setup.py clean
     598       sys.argv ['$spack-stage/spack-stage-py-ruamel-yaml-clib-0.2.0-hdspcq
             xwznokaxubih4nb37ksuocgnpr/spack-src/setup.py', 'clean', '--all']
     599       test compiling test_ruamel_yaml
  >> 600     $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-setuptools-67.6.0
             -dytmea7woljpe5ma24z6of5lhwfe36r6/lib/python3.10/site-packages/setuptools/dist.py:287: SetuptoolsDepr
             ecationWarning: The namespace_packages parameter is deprecated, consider using implicit namespaces in
             stead (PEP 420). See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-
             packages
     601         warnings.warn(msg, SetuptoolsDeprecationWarning)
     602       running clean
     603       removing 'build/temp.linux-x86_64-cpython-310' (and everything under it)
     604       removing 'build/lib.linux-x86_64-cpython-310' (and everything under it)
     605       'build/bdist.linux-x86_64' does not exist -- can't clean it
     606       'build/scripts-3.10' does not exist -- can't clean it

     ...

     609     Installing collected packages: ruamel.yaml.clib
     610       Created temporary directory: /tmp/pip-record-lm0eni1w
     611       Running setup.py install for ruamel.yaml.clib: started
     612       Running command Running setup.py install for ruamel.yaml.clib
     613       sys.argv ['$spack-stage/spack-stage-py-ruamel-yaml-clib-0.2.0-hdspcq
             xwznokaxubih4nb37ksuocgnpr/spack-src/setup.py', 'install', '--record', '/tmp/pip-record-lm0eni1w/inst
             all-record.txt', '--single-version-externally-managed', '--prefix', '$
             spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-ruamel-yaml-clib-0.2.0-hdspcqxwznokaxubih4nb37ksu
             ocgnpr', '--compile', '--install-headers', '$spack/opt/spack/linux-fed
             ora37-x86_64/gcc-12.3.1/py-ruamel-yaml-clib-0.2.0-hdspcqxwznokaxubih4nb37ksuocgnpr/include/python3.10
             /ruamel.yaml.clib']
     614       test compiling test_ruamel_yaml
  >> 615     $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-setuptools-67.6.0
             -dytmea7woljpe5ma24z6of5lhwfe36r6/lib/python3.10/site-packages/setuptools/dist.py:287: SetuptoolsDepr
             ecationWarning: The namespace_packages parameter is deprecated, consider using implicit namespaces in
             stead (PEP 420). See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-
             packages
     616         warnings.warn(msg, SetuptoolsDeprecationWarning)
     617       running install
  >> 618       $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-setuptools-67.6
             .0-dytmea7woljpe5ma24z6of5lhwfe36r6/lib/python3.10/site-packages/setuptools/command/install.py:34: Se
             tuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-bas
             ed tools.
     619         warnings.warn(
     620       running build
     621       running build_py
     622       creating build
     623       creating build/lib.linux-x86_64-cpython-310
     624       creating build/lib.linux-x86_64-cpython-310/ruamel

     ...

     1029            |                                                                                           |
     1030            |                                                                                           c
             har *
     1031      yaml.h:636:43: note: expected 'yaml_char_t *' {aka 'unsigned char *'} but argument is of type 'char
              *'
     1032        636 |         yaml_char_t *anchor, yaml_char_t *tag, int implicit,
     1033            |                              ~~~~~~~~~~~~~^~~
     1034      _ruamel_yaml.c: In function '__pyx_tp_dealloc_12_ruamel_yaml_CParser':
  >> 1035    _ruamel_yaml.c:23880:5: error: lvalue required as increment operand
     1036      23880 |     ++Py_REFCNT(o);
     1037            |     ^~
  >> 1038      _ruamel_yaml.c:23882:5: error: lvalue required as decrement operand
     1039      23882 |     --Py_REFCNT(o);
     1040            |     ^~
     1041      _ruamel_yaml.c: In function '__pyx_tp_dealloc_12_ruamel_yaml_CEmitter':
  >> 1042      _ruamel_yaml.c:24053:5: error: lvalue required as increment operand
     1043      24053 |     ++Py_REFCNT(o);
     1044            |     ^~
  >> 1045      _ruamel_yaml.c:24055:5: error: lvalue required as decrement operand
     1046      24055 |     --Py_REFCNT(o);
     1047            |     ^~
     1048      _ruamel_yaml.c: In function '__Pyx_ParseOptionalKeywords':
     1049      _ruamel_yaml.c:26224:21: warning: '_PyUnicode_get_wstr_length' is deprecated [-Wdeprecated-declarat
             ions]
     1050      26224 |                     (PyUnicode_GET_SIZE(**name) != PyUnicode_GET_SIZE(key)) ? 1 :
     1051            |                     ^

     ...

     1162          setup_py_code = "from setuptools import setup; setup()"
     1163    
     1164      exec(compile(setup_py_code, filename, "exec"))
     1165      '"'"''"'"''"'"' % ('"'"'$spack-stage/spack-stage-py-ruamel-yaml-clib
             -0.2.0-hdspcqxwznokaxubih4nb37ksuocgnpr/spack-src/setup.py'"'"',), "<pip-setuptools-caller>", "exec")
             )' install --record /tmp/pip-record-lm0eni1w/install-record.txt --single-version-externally-managed -
             -prefix $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-ruamel-ya
             ml-clib-0.2.0-hdspcqxwznokaxubih4nb37ksuocgnpr --compile --install-headers $spack/opt/spack/linux-fed
			 ora37-x86_64/gcc-12.3.1/py-ruamel-yaml-clib-0.2.0-hdspcqxwznokaxubih4n
             b37ksuocgnpr/include/python3.10/ruamel.yaml.clib
     1166      cwd: $spack-stage/spack-stage-py-ruamel-yaml-clib-0.2.0-hdspcqxwznok
             axubih4nb37ksuocgnpr/spack-src/
     1167      Running setup.py install for ruamel.yaml.clib: finished with status 'error'
  >> 1168    error: legacy-install-failure
     1169    
     1170    × Encountered error while trying to install package.
     1171    ╰─> ruamel.yaml.clib
     1172    
     1173    note: This is an issue with the package mentioned above, not pip.
     1174    hint: See above for output from the failure.
```
</details>

Then in this PR here for `pd-dm-tree`, previously it also failed with `legacy-install-failure` but changing the python version did not help (and also did not make sense).
Looking via google for the error gives a lot of result like update pip or update setuptools, that's why I first thought updating pip would help, while in reality `legacy-install-failure` is just misleading.
In this PR the actual problem was the missing `cmake` build dependency which you only see if you look deeper in spack-build-out.txt where you somewhere find the line
```
  FileNotFoundError: [Errno 2] No such file or directory: 'cmake'
```

<details>
  <summary>Full error message of `py-dm-tree@0.1.8` without cmake build dep</summary>

```
==> Installing py-dm-tree-0.1.8-r5fj576k6axcyo3lsnsqssdyea3eels6 [25/25]
==> No binary for py-dm-tree-0.1.8-r5fj576k6axcyo3lsnsqssdyea3eels6 found: installing from source
==> Using cached archive: $spack/var/spack/cache/_source-cache/archive/0f/0fcaabbb14e7980377439e7140bd05552739ca5e515ecb3119f234acee4b9430.tar.gz
==> Ran patch() for py-dm-tree
==> py-dm-tree: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/python-3.10.10-hv6jvlmupid2ylf5gt2m6ahtojappsuf/bin/python3.10' '-m' 'pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-dm-tree-0.1.8-r5fj576k6axcyo3lsnsqssdyea3eels6' '.'

2 errors found in build log:
     158    Failed to build dm-tree
     159    Installing collected packages: dm-tree
     160      Created temporary directory: /tmp/pip-record-h5fdtq1h
     161      Running setup.py install for dm-tree: started
     162      Running command Running setup.py install for dm-tree
     163      running install
  >> 164    $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-setuptools-67.6.0-
            dytmea7woljpe5ma24z6of5lhwfe36r6/lib/python3.10/site-packages/setuptools/command/install.py:34: Setupt
            oolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based to
            ols.
     165        warnings.warn(
     166      running build
     167      running build_py
     168      creating build
     169      creating build/lib.linux-x86_64-cpython-310
     170      creating build/lib.linux-x86_64-cpython-310/tree

     ...

     270          setup_py_code = "from setuptools import setup; setup()"
     271    
     272      exec(compile(setup_py_code, filename, "exec"))
     273      '"'"''"'"''"'"' % ('"'"'$spack-stage/spack-stage-py-dm-tree-0.1.8-r5f
            j576k6axcyo3lsnsqssdyea3eels6/spack-src/setup.py'"'"',), "<pip-setuptools-caller>", "exec"))' install 
            --record /tmp/pip-record-h5fdtq1h/install-record.txt --single-version-externally-managed --prefix
			$spack/linux-fedora37-x86_64/gcc-12.3.1/py-dm-tree-0.1.8-r5fj576k6axcyo3lsnsqssdyea3eels6 --compile 
			--install-headers $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-dm-tree-0.1.8-
			r5fj576k6axcyo3lsnsqssdyea3eels6/include/python3.10/d
            m-tree
     274      cwd: $spack-stage/spack-stage-py-dm-tree-0.1.8-r5fj576k6axcyo3lsnsqss
            dyea3eels6/spack-src/
     275      Running setup.py install for dm-tree: finished with status 'error'
  >> 276    error: legacy-install-failure
     277    
     278    × Encountered error while trying to install package.
     279    ╰─> dm-tree
     280    
     281    note: This is an issue with the package mentioned above, not pip.
     282    hint: See above for output from the failure.
```
</details>

I had the same problems with py-protobuf in #38614, where the `legacy-install-failure` totally confused me. I got it because `py-protobuf@4.23.3` does not support `+cpp` anymore

<details>
  <summary>Full error message of `py-protobuf@4.23.3` without disabled cpp variant</summary>

```
==> Installing py-protobuf-4.23.3-pcbvxlkg4abxyv6g5yk7pd6hpm575p4e [27/27]
==> No binary for py-protobuf-4.23.3-pcbvxlkg4abxyv6g5yk7pd6hpm575p4e found: installing from source
==> Using cached archive: $spack/var/spack/cache/_source-cache/archive/7a/7a92beb30600332a52cdadbedb40d33fd7c8a0d7f549c440347bc606fb3fe34b.tar.gz
==> No patches needed for py-protobuf
==> py-protobuf: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/python-3.10.10-hv6jvlmupid2ylf5gt2m6ahtojappsuf/bin/python3.10' '-m' 'pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-protobuf-4.23.3-pcbvxlkg4abxyv6g5yk7pd6hpm575p4e' '--install-option=--cpp_implementation' '.'

3 errors found in build log:
     15    Processing $spack-stage/spack-stage-py-protobuf-4.23.3-pcbvxlkg4abxyv6g5
           yk7pd6hpm575p4e/spack-src
     16      Added file://$spack-stage/spack-stage-py-protobuf-4.23.3-pcbvxlkg4abxy
           v6g5yk7pd6hpm575p4e/spack-src to build tracker '/tmp/pip-build-tracker-raol0991'
     17      Running setup.py (path:$spack-stage/spack-stage-py-protobuf-4.23.3-pcb
           vxlkg4abxyv6g5yk7pd6hpm575p4e/spack-src/setup.py) egg_info for package from file://$spack-stage/spack-
		   stage-py-protobuf-4.23.3-pcbvxlkg4abxyv6g5yk7pd6hpm575p4e/spack-src
     18      Created temporary directory: /tmp/pip-pip-egg-info-_w5svb1r
     19      Preparing metadata (setup.py): started
     20      Running command python setup.py egg_info
  >> 21    $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-setuptools-67.6.0-d
           ytmea7woljpe5ma24z6of5lhwfe36r6/lib/python3.10/site-packages/setuptools/dist.py:287: SetuptoolsDeprecat
           ionWarning: The namespace_packages parameter is deprecated, consider using implicit namespaces instead 
           (PEP 420). See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
     22        warnings.warn(msg, SetuptoolsDeprecationWarning)
     23      running egg_info
     24      creating /tmp/pip-pip-egg-info-_w5svb1r/protobuf.egg-info
     25      writing /tmp/pip-pip-egg-info-_w5svb1r/protobuf.egg-info/PKG-INFO
     26      writing dependency_links to /tmp/pip-pip-egg-info-_w5svb1r/protobuf.egg-info/dependency_links.txt
     27      writing namespace_packages to /tmp/pip-pip-egg-info-_w5svb1r/protobuf.egg-info/namespace_packages.txt

     ...

     36    Created temporary directory: /tmp/pip-unpack-pw0z0ye9
     37    Installing collected packages: protobuf
     38      DEPRECATION: protobuf is being installed using the legacy 'setup.py install' method, because the '--n
           o-binary' option was enabled for it and this currently disables local wheel building for projects that 
           don't have a 'pyproject.toml' file. pip 23.1 will enforce this behaviour change. A possible replacement
            is to enable the '--use-pep517' option. Discussion can be found at https://github.com/pypa/pip/issues/
           11451
     39      Created temporary directory: /tmp/pip-record-05fk9kd6
     40      Running setup.py install for protobuf: started
     41      Running command Running setup.py install for protobuf
  >> 42    $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-setuptools-67.6.0-d
           ytmea7woljpe5ma24z6of5lhwfe36r6/lib/python3.10/site-packages/setuptools/dist.py:287: SetuptoolsDeprecat
           ionWarning: The namespace_packages parameter is deprecated, consider using implicit namespaces instead 
           (PEP 420). See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
     43        warnings.warn(msg, SetuptoolsDeprecationWarning)
     44      usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     45         or: setup.py --help [cmd1 cmd2 ...]
     46         or: setup.py --help-commands
     47         or: setup.py cmd --help
     48    

     ...

     89          setup_py_code = "from setuptools import setup; setup()"
     90    
     91      exec(compile(setup_py_code, filename, "exec"))
     92      '"'"''"'"''"'"' % ('"'"'$spack-stage/spack-stage-py-protobuf-4.23.3-pc
           bvxlkg4abxyv6g5yk7pd6hpm575p4e/spack-src/setup.py'"'"',), "<pip-setuptools-caller>", "exec"))' install 
           --record /tmp/pip-record-05fk9kd6/install-record.txt --single-version-externally-managed --prefix $spa
		   ck/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-protobuf-4.23.3-pcbvxlkg4abxyv6g5yk7pd6hpm575p4e 
		   --compile --install-headers $spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/py-protobuf-4.23.3
		   -pcbvxlkg4abxyv6g5yk7pd6hpm575p4e/include/python3.10/p
           rotobuf --cpp_implementation
     93      cwd: $spack-stage/spack-stage-py-protobuf-4.23.3-pcbvxlkg4abxyv6g5yk7p
           d6hpm575p4e/spack-src/
     94      Running setup.py install for protobuf: finished with status 'error'
  >> 95    error: legacy-install-failure
     96    
     97    × Encountered error while trying to install package.
     98    ╰─> protobuf
     99    
     100   note: This is an issue with the package mentioned above, not pip.
     101   hint: See above for output from the failure.
```
</details>

Here again, when looking deeper in spack-build.log.txt there is somewhere the line 
```
  error: option --cpp_implementation not recognized
```
which is actually the cause of the problem.

So to summarize:
- from not having seen `legacy-install-failure` before, I encountered it multiple times now in 2 days
- it's often just a red herring and you have to look deeper in the build log to see what the actual problem is
- the reason why I have asked if something has changed: I believe to remember that in the past building just failed with the actual error and I don't think I have seen `legacy-install-failure` before

File: var/spack/repos/builtin/packages/py-dm-tree/package.py (hunk 1 above)
Comment (thomas-bouvier) [reply var/spack/repos/builtin/packages/py-dm-tree/package.py]: My bad, I got confused. I fixed it!

**Review by manuelakuhn — APPROVED** [review]

**Comment by adamjstewart** [comment]
I'm guessing "legacy install failure" just means the package was built using a legacy install method (in this case `setup.py` instead of `pyproject.toml`) and there was a failure during install. Spack's automatic error parsing from logs has always been useless for Python, it's based on CMake instead. If someone has the energy it would be good to improve that, and I can point you to where. But for now, I pretty much ignore the errors that show up on the command line and look at the log instead.

**Review by adamjstewart — APPROVED** [review]
Excited to see a package switch from Bazel to CMake!

**Comment by adamjstewart** [comment]
Will give @aweits a chance to review this, ping me if it's been a week and you haven't heard anything.

**Comment by thomas-bouvier** [comment]
What are the strenghts of CMake over Bazel, @adamjstewart?

**Comment by adamjstewart** [comment]
I'm far from an expert on this, but in my experience with Spack, CMake packages "Just Work". Bazel packages tend to be tightly locked to a specific version of Bazel, and tend not to work unless you're on the exact system that Google claims to support. I've also found Bazel builds to be very slow despite that being the main reason Google made Bazel. Bazel likes to download and manage all of its dependencies, making it difficult to integrate into package managers where each dependency should be managed by Spack instead. Most dependencies are pinned to a random commit and don't work with the stable release before or after. And if one of those URLs or checksums changes, you have to patch Bazel config files just to get them to work. If you look at the 1K+ line monstrosity that is our TensorFlow recipe, you'll get an idea of my frustrations with Bazel-based packages.

**Comment by thomas-bouvier** [comment]
> I'm far from an expert on this, but in my experience with Spack, CMake packages "Just Work". Bazel packages tend to be tightly locked to a specific version of Bazel, and tend not to work unless you're on the exact system that Google claims to support. I've also found Bazel builds to be very slow despite that being the main reason Google made Bazel. Bazel likes to download and manage all of its dependencies, making it difficult to integrate into package managers where each dependency should be managed by Spack instead. Most dependencies are pinned to a random commit and don't work with the stable release before or after. And if one of those URLs or checksums changes, you have to patch Bazel config files just to get them to work. If you look at the 1K+ line monstrosity that is our TensorFlow recipe, you'll get an idea of my frustrations with Bazel-based packages.

Thank you for the detailed answer!
